File: Python/ilinkai_weixin/media/media_download_service.py
"""媒体下载服务"""
import logging
import os
import uuid
from datetime import datetime

from ..cdn.cdn_client import CdnClient

logger = logging.getLogger(__name__)
MAX_MEDIA_BYTES = 100 * 1024 * 1024


class MediaDownloadService:
    def __init__(self, cdn_base_url: str):
        self._cdn_base_url = cdn_base_url
        self._cdn_client = CdnClient()

    def download_image(self, encrypt_query_param: str, aes_key_base64: str = None) -> bytes:
        if not encrypt_query_param:
            raise ValueError("encrypt_query_param is required")
        if aes_key_base64:
            return self._cdn_client.download_and_decrypt(
                encrypt_query_param, aes_key_base64, self._cdn_base_url, "DownloadImage")
        return self._cdn_client.download_plain(
            encrypt_query_param, self._cdn_base_url, "DownloadImage-plain")

    def download_voice(self, encrypt_query_param: str, aes_key_base64: str) -> bytes:
        if not encrypt_query_param or not aes_key_base64:
            raise ValueError("encrypt_query_param and aes_key_base64 are required")
        return self._cdn_client.download_and_decrypt(
            encrypt_query_param, aes_key_base64, self._cdn_base_url, "DownloadVoice")

    def download_file(self, encrypt_query_param: str, aes_key_base64: str) -> bytes:
        if not encrypt_query_param or not aes_key_base64:
            raise ValueError("encrypt_query_param and aes_key_base64 are required")
        return self._cdn_client.download_and_decrypt(
            encrypt_query_param, aes_key_base64, self._cdn_base_url, "DownloadFile")

    def download_video(self, encrypt_query_param: str, aes_key_base64: str) -> bytes:
        if not encrypt_query_param or not aes_key_base64:
            raise ValueError("encrypt_query_param and aes_key_base64 are required")
        return self._cdn_client.download_and_decrypt(
            encrypt_query_param, aes_key_base64, self._cdn_base_url, "DownloadVideo")

    @staticmethod
    def save_media_to_file(data: bytes, directory: str, file_name: str = None) -> str:
        os.makedirs(directory, exist_ok=True)
        if not file_name:
            file_name = f"media_{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}_{uuid.uuid4().hex}.bin"
        file_path = os.path.join(directory, file_name)
        with open(file_path, 'wb') as f:
            f.write(data)
        logger.debug("SaveMediaToFile: saved to %s", file_path)
        return file_path

File: Java/src/main/java/com/ilinkai/weixin/media/MediaDownloadService.java
package com.ilinkai.weixin.media;

import com.ilinkai.weixin.cdn.CdnClient;
import com.ilinkai.weixin.cdn.CdnException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.UUID;

/** 媒体下载服务 */
public class MediaDownloadService {
    private static final Logger logger = LoggerFactory.getLogger(MediaDownloadService.class);
    public static final int MAX_MEDIA_BYTES = 100 * 1024 * 1024;

    private final CdnClient cdnClient;
    private final String cdnBaseUrl;

    public MediaDownloadService(String cdnBaseUrl) {
        this.cdnBaseUrl = cdnBaseUrl;
        this.cdnClient = new CdnClient();
    }

    public byte[] downloadImage(String encryptQueryParam, String aesKeyBase64) throws CdnException {
        if (encryptQueryParam == null || encryptQueryParam.isEmpty())
            throw new IllegalArgumentException("encryptQueryParam is required");
        if (aesKeyBase64 != null && !aesKeyBase64.isEmpty()) {
            return cdnClient.downloadAndDecrypt(encryptQueryParam, aesKeyBase64, cdnBaseUrl, "DownloadImage");
        } else {
            return cdnClient.downloadPlain(encryptQueryParam, cdnBaseUrl, "DownloadImage-plain");
        }
    }

    public byte[] downloadVoice(String encryptQueryParam, String aesKeyBase64) throws CdnException {
        if (encryptQueryParam == null || aesKeyBase64 == null || encryptQueryParam.isEmpty() || aesKeyBase64.isEmpty())
            throw new IllegalArgumentException("encryptQueryParam and aesKeyBase64 are required");
        return cdnClient.downloadAndDecrypt(encryptQueryParam, aesKeyBase64, cdnBaseUrl, "DownloadVoice");
    }

    public byte[] downloadFile(String encryptQueryParam, String aesKeyBase64) throws CdnException {
        if (encryptQueryParam == null || aesKeyBase64 == null || encryptQueryParam.isEmpty() || aesKeyBase64.isEmpty())
            throw new IllegalArgumentException("encryptQueryParam and aesKeyBase64 are required");
        return cdnClient.downloadAndDecrypt(encryptQueryParam, aesKeyBase64, cdnBaseUrl, "DownloadFile");
    }

    public byte[] downloadVideo(String encryptQueryParam, String aesKeyBase64) throws CdnException {
        if (encryptQueryParam == null || aesKeyBase64 == null || encryptQueryParam.isEmpty() || aesKeyBase64.isEmpty())
            throw new IllegalArgumentException("encryptQueryParam and aesKeyBase64 are required");
        return cdnClient.downloadAndDecrypt(encryptQueryParam, aesKeyBase64, cdnBaseUrl, "DownloadVideo");
    }

    public String saveMediaToFile(byte[] data, String directory, String fileName) throws IOException {
        new File(directory).mkdirs();
        if (fileName == null || fileName.isEmpty()) {
            fileName = "media_" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"))
                    + "_" + UUID.randomUUID().toString().replace("-", "") + ".bin";
        }
        String filePath = directory + File.separator + fileName;
        try (FileOutputStream fos = new FileOutputStream(filePath)) {
            fos.write(data);
        }
        logger.debug("SaveMediaToFile: saved to {}", filePath);
        return filePath;
    }
}
